h5py/h5t.pyx

```diff
@@ -266,6 +269,14 @@ H5Tlock(H5PY_OBJ)
 
 PYTHON_OBJECT = lockid(H5PY_OBJ)
 
+# HDF5 datatype to represent Numpy (>=2) variable-length strings
+# Should never be directly stored in files.
+cdef hid_t H5PY_NPYSTR = H5Tcreate(H5T_OPAQUE, 2 * sizeof(size_t))
```

Comment:
Just to note, the size I specified here kind of relies on the in-memory format of the numpy string datatype, which is [described in NEP 55](https://numpy.org/neps/nep-0055-string_dtype.html#memory-layout-and-managing-heap-allocations), but it's not part of the API. I don't *think* it should actually matter if it's wrong, since we're just using the datatype as a marker, and handling all the conversion ourselves, but I haven't checked that.

I guess there's an argument that we should set it to a totally incorrect value, like 1 byte, so that nothing can rely on the size. :shrug: